### ISSUE TITLE
Compress sparsity pattern column data

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#11356]
I've been listening to talks by the Exascale Computing Project for the past couple of days, and it's giving me lots of ideas for stuff one could try out.

In sparse matrices, about 1/3 of the memory is for storing the column indices (stored in the `SparsityPattern` class). I suspect that that could relatively easily be compressed. One approach is the `ChumkSparsityPattern` class. Another would be as follows: Instead of storing column indices, one could store deltas and in many cases these will actually be quite small: I would expect that in many cases, these could actually be stores in just one byte instead of the four bytes we currently use. It wouldn't be very difficult to compress this then: Use 1 byte if possible, if not use four bytes. In the best of cases, this would reduce memory usage to 9 instead of 12 bytes per matrix entry, and one could imagine that because memory traffic is the limitation for many solvers, this would increase performance by 30%.

**Comment by kronbichler** [comment]
> one could imagine that because memory traffic is the limitation for many solvers, this would increase performance by 30%.

This is definitely true and makes a lot of sense. It would make for a nice student project. On the other hand, if you care about performance, you can use matrix-free solvers and, as long as the preconditioner toolchain fits, you don't just get 30% but rather 3x or more speedup. :wink:

**Comment by bangerth** [comment]
Absolutely! But it's also a lot of work to convert, and a lot of people do use matrix-based. There's no reason *not* to make this case fast *as well* :-)
